### ISSUE TITLE
MPV exits immediately when watching a specific video with ao=pipewire

### ISSUE THREAD

**Issue by CoelacanthusHex** [mpv-player/mpv#10545]
### Important Information

Provide following Information:
- mpv https://github.com/mpv-player/mpv/commit/891c9c869f61bfe16e46a9244a6af2ed93f67b23
- Linux Distribution and Version: Arch
- Source of the mpv binary: https://github.com/mpv-player/mpv/commit/891c9c869f61bfe16e46a9244a6af2ed93f67b23
- If known which version of mpv introduced the problem: https://github.com/mpv-player/mpv/commit/dbfee1be3d6da98d11517e1d166216c715720711
- Window Manager and version: KDE 5.25.4

If you're not using git master or the latest release, update.
Releases are listed here: https://github.com/mpv-player/mpv/releases

### Reproduction steps

use mpv after https://github.com/mpv-player/mpv/commit/dbfee1be3d6da98d11517e1d166216c715720711
Download https://mirrors.tuna.tsinghua.edu.cn/tuna/tunight/2021-06-12-homemade-vm/video.mp4
mpv video.mp4 --no-config --ao=pipewire

### Expected behavior

Play normally

### Actual behavior

exit immediately

### Log file

https://0x0.st/oLmL.txt

The issue will be closed for ignoring the issue template.

### Sample files

https://mirrors.tuna.tsinghua.edu.cn/tuna/tunight/2021-06-12-homemade-vm/video.mp4

https://0x0.st/oLmp.mp4

**Comment by christoph-heinrich** [comment]
Is that the full log? It ends at `[vd] Allocating new DR image...` ?
I've tried reproducing it on the same commit, using the same OS, with the same library versions, on the same desktop environment (on wayland), and also with an AMD GPU and the command you provided works without a problem for me.
My output is pretty much the same, except that it continues afterwards.

**Comment by CoelacanthusHex** [comment]
> Is that the full log? It ends at `[vd] Allocating new DR image...`

Yes, that is all. `[vd] Allocating new DR image...` will appear between the last serval lines.

> ? I've tried reproducing it on the same commit, using the same OS, with the same library versions, on the same desktop environment (on wayland), and also with an AMD GPU and the command you provided works without a problem for me. My output is pretty much the same, except that it continues afterwards.

I found it actually is SIGSEGV, at https://github.com/mpv-player/mpv/blob/891c9c869f61bfe16e46a9244a6af2ed93f67b23/audio/out/ao_pipewire.c#L132
it got buf->datas = 0, so it crashed (actually all member of buff is zero)
and when I debug it with gdb, it didn't crash when first run this line, so I think set nframes = b->requested may change some state then cause the wrong value.

> 
> Regarding the title: your problem might be that pipewire is an ao and not a vo 😆

fixed

**Comment by CoelacanthusHex** [comment]
And since it depends on pipewire behavior, I will try again without my pipewire config.

**Comment by CoelacanthusHex** [comment]
Some bad software generate wrong config in my config directory, so it's not mpv's bad.
the wrong config is
```json
{ 
    "restore.stream.Output/Audio.application.name:mpv": { "volume": 1.000000, "mute": false, "volumes": [ 1.000000, 1.000000 ], "channels": [ "FL", "FR" ] }
}
```
in `~/.config/pipewire/media-session.d/restore-stream`

**Comment by sfan5** [comment]
A wrong pipewire config should not make mpv segfault, however.

**Comment by philipl** [comment]
@t-8ch Can you take a look?

**Comment by t-8ch** [comment]
@CoelacanthusHex Can you try to provide a full stacktrace for the segfault? Something doesn't add up here.

**Comment by t-8ch** [comment]
Also could you explain why you think the pipewire config you posted is incorrect and if the video is supposed to actually play sound. (It does not for me)

**Comment by t-8ch** [comment]
Ok, the sound and video only starts later. I was also able to reproduce the crash when seeking, but it's not a stable reproducer.

**Comment by t-8ch** [comment]
This looks like a bug in Pipewire to me. `b->requested == 75` but `b->n_datas == 0`. So Pipewire asks for samples but does not provide any buffers to store those samples.

@CoelacanthusHex can you try this patch:
```diff
diff --git a/audio/out/ao_pipewire.c b/audio/out/ao_pipewire.c
index 3421856c44..d46870679b 100644
--- a/audio/out/ao_pipewire.c
+++ b/audio/out/ao_pipewire.c
@@ -129,6 +129,12 @@ static void on_process(void *userdata)

     struct spa_buffer *buf = b->buffer;

+    if (buf->n_datas == 0) {
+        MP_WARN(ao, "can not write data to nonexistent buffer\n");
+        pw_stream_queue_buffer(p->stream, b);
+        return;
+    }
+
     int bytes_per_channel = buf->datas[0].maxsize / ao->channels.num;
     int nframes = bytes_per_channel / ao->sstride;
 #if PW_CHECK_VERSION(0, 3, 49)
```

Maybe @wtay can chime in if this is expected.

**Comment by t-8ch** [comment]
Also logs with `PIPEWIRE_DEBUG=9` would help.

**Comment by CoelacanthusHex** [comment]
> This looks like a bug in Pipewire to me. `b->requested == 75` but `b->n_datas == 0`. So Pipewire asks for samples but does not provide any buffers to store those samples.
> 
> @CoelacanthusHex can you try this patch:
> 
> ```diff
> diff --git a/audio/out/ao_pipewire.c b/audio/out/ao_pipewire.c
> index 3421856c44..d46870679b 100644
> --- a/audio/out/ao_pipewire.c
> +++ b/audio/out/ao_pipewire.c
> @@ -129,6 +129,12 @@ static void on_process(void *userdata)
> 
>      struct spa_buffer *buf = b->buffer;
> 
> +    if (buf->n_datas == 0) {
> +        MP_WARN(ao, "can not write data to nonexistent buffer\n");
> +        pw_stream_queue_buffer(p->stream, b);
> +        return;
> +    }
> +
>      int bytes_per_channel = buf->datas[0].maxsize / ao->channels.num;
>      int nframes = bytes_per_channel / ao->sstride;
>  #if PW_CHECK_VERSION(0, 3, 49)
> ```
> 

It works.

**Comment by t-8ch** [comment]
@CoelacanthusHex Thanks! Could you also provide the output of the following command:
```
PIPEWIRE_DEBUG=9 mpv --msg-level=all=debug --no-config --ao=pipewire video.mp4
```

It would be useful to know why we get the incorrect request from Pipewire, so we can fix it properly.

**Comment by CoelacanthusHex** [comment]
> @CoelacanthusHex Thanks! Could you also provide the output of the following command:
> 
> ```
> PIPEWIRE_DEBUG=9 mpv --msg-level=all=debug --no-config --ao=pipewire video.mp4
> ```
> 
> It would be useful to know why we get the incorrect request from Pipewire, so we can fix it properly.

https://0x0.st/oLxK.txt

**Comment by t-8ch** [comment]
@CoelacanthusHex Thanks, but it seems this dump does not actually exhibit the issue anymore.
I was also not able to reproduce it after the first time it happened.

We can add the warning so mpv doesn't crash and wait for somebody else to reproduce it.

**Comment by t-8ch** [comment]
https://github.com/mpv-player/mpv/pull/10559 could be a fix for this.

**Comment by thebombzen** [comment]
@CoelacanthusHex can you reproduce this anymore?

**Comment by CoelacanthusHex** [comment]
> @CoelacanthusHex can you reproduce this anymore?

I can't reproduce on HEAD.

**Comment by thebombzen** [comment]
Fixed by 6c1f01d
